src/pip/_internal/req/__init__.py

```diff
@@ -60,39 +71,95 @@ def install_given_reqs(
             ', '.join(to_install.keys()),
         )
 
-    installed = []
+    # pre allocate installed package names
+    installed = collections.OrderedDict({name: None for name in to_install})
+
+    install_args = [install_options, global_options, dict(
+        root=root, home=home, prefix=prefix,
+        warn_script_location=warn_script_location,
+        use_user_site=use_user_site, pycompile=pycompile)]
```

Comment:
Can we make this a namedtuple instead? The `install_args[0], install_args[1], **install_args[2]` call below gives me goosebumps.

It’s probably also a good idea to introduce some type hints on the kwarg dict with either TypeDict or nameduple.

Reply:
namedtuple sounds good, regrading TypeDict on the kwargs, I'd prefer to avoid adding additional typing as it really just passes the values. Reasoning is that if someone decides to change typing on the original function, they have to go through another search/replace with little upside/visibility, wdyt?

Reply:
Some prefer TypedDict (which avoids adding a concrete type) and some namedtuple (which avoids adding typing). I’m fine with either, so feel free to go what you feel is better.